### ISSUE TITLE
Problem: Too many config notations

### ISSUE THREAD

**Issue by maennchen** [c-rack/quantum-elixir#174]
# Notations

Do we really need that many config notations. I think they are confusing.

I propose to reduce the configuration options to following:

## Supported Schedule Notations in mix config (with #156 in mind)

Those notations are only allowed, because the sigil doesn't work in config.

1. `{:cron, "* * * * *"}` (Tuple of `:cron`, String) - Expression without second support
2. `{:extended, "* * * * * *"}` (Tuple of `:extended`, String) - Expression with second support
3. `"* * * * *"` (Any String) - Expression without second support

## Supported Schedule Notations in Runtime

Config Notations are not supported.

If string is provided, raise error with details on how to get a `Crontab.CronExpression` struct.

1. `~e[* * * *]` - Only `Crontab.CronExpression` Struct is Supported (encouraged via Sigil)

## Supported Task Notations in mix config

1.  ~~`&Backup.backup/0` - Standard Way of specifying Tasks~~
2. ~~`{"Heartbeat", :send}` - Important because 1 wouldn't work with tools like `distillery`~~
3. `fn -> :some_anon_func end}` - Used for simple short tasks
4. `{Module, :func, [args]}` - Standard Way of specifying Tasks

## Supported Task Notations in Runtime
1.  ~~`&Backup.backup/0` - Standard Way of specifying Tasks~~
2. `fn -> :some_anon_func end}` - Used for simple short tasks
3. `{Module, :func, [args]}` - Standard Way of specifying Tasks

## Dropped Formats

1. `"* * * * * MyApp.MyModule.my_method"` - Not needed in my opinion, can be solved with Schedule Notation 3.
2. `&Backup.backup/0` - Use `{Module, :func, [args]}` instead
4. `{"Heartbeat", :send}` - Use `{Module, :func, [args]}` instead
4. `{Heartbeat, :send}` - Use `{Module, :func, [args]}` instead

## Mix Configuration Syntax

```elixir
config :quantum, :your_app,
  cron: [
    # Named Explicit Form
   [name: NAME, schedule: CONFIG_SCHEDULE_NOTATION, task: CONFIG_TASK_NOTATION, OTHER_FIELDS],

    # Unnamed Explicit Form
   [schedule: CONFIG_SCHEDULE_NOTATION, task: CONFIG_TASK_NOTATION, OTHER_FIELDS],
    # Short Form
   {CONFIG_SCHEDULE_NOTATION, CONFIG_TASK_NOTATION},
  ]
```

## Runtime Configuration Syntax

```elixir
# Short Form
Quantum.add_job(RUNTIME_SCHEDULE_NOTATION, RUNTIME_TASK_NOTATION)

# Explicit Form
Quantum.add_job(JOB_STRUCT)
```

# Job Struct
Instead of Normalizing the Job Struct, we provide factory functions as well as setter for them.

The setters are easier to strictly type and it should be more clear for the user on how to handle them.

Usage of the Job Struct without the factory & setters is strongly discouraged.

Example:

```elixir
alias Qauntum.Job

Quantum.new_job() # Produces Job with default values like Timezone etc.
|> Job.put_schedule(RUNTIME_SCHEDULE_NOTATION) # Normalizes Value
|> Job.put_task(RUNTIME_TASK_NOTATION) # Normalizes Value
# ....
```

Invalid values give much clearer stack traces with the setter function.

Example:

```elixir
alias Qauntum.Job

Quantum.new_job()
|> Job.put_schedule("invalid, an expression should be here")
** (FunctionClauseError) no function clause matching in Quantum.Job.put_schedule/1
```

## Removed `args`

This is replaced by the new `{Module, :func, [args]}` notation.

**Comment by michalmuskala** [comment]
I would propose using the standard `{module, function, args}` notation for encoding jobs. It's extremely widespread, works in both runtime and config, is very familiar and does not require changing strings to atoms, since the module is already an atom. The "mfargs" tuple and 0-arity anonymous functions should cover all the use cases.

To be honest, I have no idea why in the current `{module, function}` notation module is a string.

**Comment by maennchen** [comment]
@michalmuskala I'm guessing that it allows string, to allow configuration via Release Managers.

I think that that sort of thing shouldn't be configured via Release Managers. (The Schedule I can understand, but not the function.)

Therefore I'd be happy to remove `Supported Task Notations in mix config#2`.

**Comment by maennchen** [comment]
@michalmuskala I updated the issue description. Do you think it is reasonable now?

**Comment by maennchen** [comment]
What I don't like in the new notation, is that it is conflicting with `Quantum.Job.args`. It would mean, that ew remove this struct property.

**Comment by maennchen** [comment]
Closed with #178
